ENS Validation based on Google's IDN Policy algorithm

matthewlilley opened the pull request:
WIP

https://cs.chromium.org/chromium/src/components/url_formatter/idn_spoof_checker.cc

### Google IDN policy algorithm
- [x] Convert each component stored in the ACE to Unicode per UTS 46 transitional processing (ToUnicode).
- [x] If there is an error in ToUnicode conversion (e.g. contains disallowed characters, starts with a combining mark, or violates BiDi rules), punycode is displayed.
- [x] If any character is outside the union of the following sets, punycode is displayed.
-- Characters allowed in identifiers per Unicode Technical Standard 39 (UTS 39)
-- Characters suitable for identifiers in 5 Aspirational scripts per Unicode Standard Annex 31 (UAX 31) : Starting with Chrome 64, domains with these characters are turned to punycode.
- [x] If the component contains either U+0338 or U+2027, punycode is displayed.
- [x] Latin, Cyrillic or Greek characters cannot be mixed with each other
- [x]  Latin characters in the ASCII range can be mixed with characters in another script as long as it's not Greek nor Cyrillic.
- [x]  Latin characters in the ASCII range can be mixed ONLY with Chinese (Han, Bopomofo), Japanese (Kanji, Katakana, Hiragana), or Korean (Hangul, Hanja). 
- [x]  Latin characters in the ASCII range can be mixed ONLY with Chinese (Han, Bopomofo), Japanese (Kanji, Katakana, Hiragana), or Korean (Hangul, Hanja).
- [x]  Han (CJK Ideographs) can be mixed with Bopomofo
- [x]  Han can be mixed with Hiragana and Katakana
- [x]  Han can be mixed with Korean Hangul
- [x]  Han can be mixed with Korean Hangul
- [x] If two or more numbering systems (e.g. European digits + Bengali digits) are mixed, punycode is shown.
- [x] If there are any invisible characters (e.g. a sequence of the same combining mark or a sequence of Kana combining marks), punycode is shown. 
- [x] Test the label for mixed script confusable per UTS 39. If mixed script confusable is detected, show punycode.
- [x] If a hostname belongs to an non-IDN TLD(top-level-domain) such as 'com', 'net', or 'uk' and all the letters in a given label belong to a set of Cyrillic letters that look like Latin letters (e.g. Cyrillic Small Letter IE - е  ), show punycode.
- [x] If the label matches a dangerous pattern, punycode is shown.
- [ ] If the end of a hostname is identical to one of top 10k domains after removing diacritic marks and mapping each character to its spoofing skeleton (e.g. www.googlé.com with 'é' in place of 'e'), punycode is shown.  - IN PROGRESS

### Refactor
- [ ]

matthewlilley committed:
first
more tests and some cleanup

dternyak commented:
@matthewlilley Wow! Thanks for opening this. @pakaplace is working on a similar solution, but this looks a fair bit more developed.

matthewlilley commented:
@dternyak it's a start! I'll push some further updates later.

matthewlilley committed:
Added patterns for allowed, inclusion, and removed unicode sets.
Added detection for all scripts (Might not need this).
More tests.

matthewlilley commented:
@dternyak Would you mind reviewing so far?

dternyak commented:
@matthewlilley Absolutely - you've made a lot of progress so far.

HenryNguyen5 commented on src/enums.ts:
```diff
@@ -0,0 +1,3 @@
+enum RestrictionLevel {
+    HIGHLY_RESTRICTIVE,
```
Not sure if you're planning on implementing the other restriction levels, but if not, i'd suggest removing this and making the script `HIGHLY_RESTRICTIVE` as the only option

HenryNguyen5 commented on src/index.ts:
```diff
@@ -0,0 +1,121 @@
+    }
+}
+
+export class Validator implements Interface {
```
I think it'd make a more ergonomic API if a single function was exported instead, so on import you could do
```ts
import {validate} from 'ens-validation';
validate('mytext')
```

HenryNguyen5 replied:
Either exporting this as a singleton or breaking the class down into functions would work!
```ts
export const validate = (new Validator()).validate
// or
// break down into multiple functions, and just export validate
```

HenryNguyen5 replied:
I think the latter option would be better, since you could then unit test each function separately to get better code coverage and modularity in tests (unit test + integration tests)

matthewlilley replied:
Agreed, lets go with option 2.

HenryNguyen5 commented on src/regexes.ts:
```diff
@@ -0,0 +1,143 @@
+export const yi = XRegExp('\\p{Yi}')
+
+export function getScripts(string: string): object {
+    const scripts = {
```
Instead of making this large object here, you could have all of the above regexs in their own file, and import the entire file,
```ts
import * as scripts from './scripts' // a file containing all of your declared regexes above
```

HenryNguyen5 commented on src/regexes.ts:
```diff
@@ -0,0 +1,143 @@
+        yi
+    }
+    const result = {}
+    for(let script in scripts) {
```
Instead of `for in`, use 
```ts
for (const [scriptName, script] of Object.entries(scripts) {
// now you don't have to check for hasOwnProperty, and don't have to access the property since you have "script" as a value already
}
```

HenryNguyen5 replied:
Combining `Object.entries` with a `.reduce` would also be pretty clean here.

HenryNguyen5 commented on test/index.js:
```diff
@@ -0,0 +1,144 @@
+    nonAsciiLatin
+} = require('../src/regexes')
+
+describe('regexes', () => {
```
would prefer to have this in its own file so you can have seperation of concerns here

HenryNguyen5 reviewed (commented):
Hey matthewlilley! Thanks for your contribution, this is looking really great so far for a WIP. I've commented above where I thought API ergonomics and structuring can be improved, along with some looping construct stuff. IDN / string handling isn't my area of expertise but it your tests make it much easier to digest whats going on (Although, I think more unit tests would definitely help out). Moving on, I'd suggest you look at our [ts-package-boilerpalte](https://github.com/MyCryptoHQ/ts-package-boilerplate) to convert your lib into a package that can be consumed for both browsers and node users, along with having much stricter type constraints. If you need help using the boilerplate, I'd be more than happy to help out with doing so.

matthewlilley commented:
@HenryNguyen5 Thanks. I've made a bunch of changes based off your boilerplate and comments.

matthewlilley committed:
Refactor

- Export single "validate" function
- Incorporate standards from ts-package-boilerpalte
- Cleanup and simplify
- Fix most of the tests

matthewlilley commented:
@HenryNguyen5 I'm not too familiar with this test suite, but it's currently failing although I've made what I believe to be the required updates to the test files. Might need you to take a look here.

matthewlilley commented:
Most of the test are fixed now.

matthewlilley commented:
@vs77bb Not quite there yet, but close. I'm busy Today, but I'll jump back on this Tomorrow.

matthewlilley committed:
Add checks for mixed numerics, sequences of non-spacing marks, and initial blacklist

matthewlilley committed:
Add return type for mixedNumerics function

matthewlilley committed:
Simplify sequenceOfCombiningMarks function, and add return type

matthewlilley committed:
Similar to top domains function, using levenstein

vs77bb commented:
@matthewlilley Good to know. Thank you for dealing with Gitcoin Bot, we're still trying to train it to act appropriately 😄

matthewlilley committed:
SAVEPOINT - Initial tests are passing.

matthewlilley commented:
@dternyak @HenryNguyen5  Quick update. In the process of a large refactor, but lots of progress overall and this is getting very close to complete. Another afternoon and I think this'll be ready for some serious abuse. I've pushed my savepoint for now.

HenryNguyen5 commented:
Hey @matthewlilley! Sorry for the late reply. Sounds good, is there anything else you'd like me to look at?

dternyak commented:
Hey @matthewlilley!

Just wanted to check in here. You've been building out a really sweet library here, and we're excited to start cranking on it.

It looks like you left off in the midst of a large refactor. How is that going?

matthewlilley commented:
Hey @dternyak ! I'm really sorry for the late response, I've been away for the past fortnight. This was cleaned up, and pretty much complete before I left, so give me a day or two and it'll be here for your review. Again, sorry for the late reply.

dternyak commented:
No worries at all! Happy to have you back, and excited to review soon.

On Thu, Jun 14, 2018 at 10:21 AM Matthew Lilley <notifications@github.com>
wrote:

> Hey @dternyak <https://github.com/dternyak> ! I'm really sorry for the
> late response, I've been away for the past fortnight. This was cleaned up,
> and pretty much complete before I left, so give me a day or two and it'll
> be here for your review. Again, sorry for the late reply.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/MyCryptoHQ/ens-validation/pull/2#issuecomment-397334295>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AHf02UtE9_aHxl8gE4MTjkPZGm8SQZ6Dks5t8n9qgaJpZM4TiT3p>
> .
>

matthewlilley commented:
@dternyak Thanks! Can you review the latest save point? There's still some work to do on confusables and skeleton generation, but a majority of the checks are complete, and tests are passing.

dternyak commented:
@matthewlilley Awesome! Congratulations. I'll follow up shortly with feedback.

dternyak commented:
Hey @matthewlilley - just wanted to let you know that we're still reviewing, but have someone actively working on the review.

briansoule commented on src/algorithms/strings/index.ts:
```diff
@@ -0,0 +1 @@
+export { levenshtein } from './levenstein';
```
Looks like we misspelled `'./levenshtein'` here and in the `src/algorithms/strings/levenstein.ts` filename.

briansoule commented on package.json:
```diff
@@ -0,0 +1,98 @@
+{
+  "name": "ens-validation",
+  "version": "0.0.0",
+  "description": "",
```
Are we holding off on the npm description?

briansoule replied:
Also adding `"keywords": ["ens"],` might help people find us.

briansoule commented on README.md:
```diff
@@ -1 +1,15 @@
-# ens-validation
+# ENS Validation
+
+Following the alorithm used for Google's IDN policy, this library can validate a ENS domain.
```
I could probably add a little color here if it would help

briansoule commented on README.md:
````diff
@@ -1 +1,15 @@
+
+```sh
+yarn build
+```
````
Perhaps we could add a little section about how to contribute additional examples to domains.json.

briansoule reviewed (approved):
Damn good stuff guys.

dternyak commented:
@matthewlilley thank you so much for your work on this. We're going to move forward with merging, and can make touchups as needed in the future.

We're going to get that bounty unlocked and reach out for some more discussion about the bounty :)

alihaiderra152-commits reviewed (commented):
Thanks so much guys I salute you all for your constant effort to make this  all work .you guys are the reason this world is a better place.and also name service I meant google
